src/views/virtualmachines/actions/utils.ts

```diff
@@ -0,0 +1,95 @@
+import { getDataVolumeTemplates, getVolumes } from '@kubevirt-utils/resources/vm';
+import { k8sDelete, Patch } from '@openshift-console/dynamic-plugin-sdk';
+
+import { ANNOTATION_PREFIX_MIGRATION_ORIGIN_CLAIMNAME } from './constants';
+
+export const deleteUnusedDataVolumes = (vm: V1VirtualMachine, namespace: string) => {
+  return getVolumes(vm)
+    ?.map((volume) => volume.dataVolume?.name)
+    ?.map((claimName) =>
+      k8sDelete({
+        model: DataVolumeModel,
+        resource: {
+          apiVersion: `${DataVolumeModel.apiGroup}/${DataVolumeModel.apiVersion}`,
+          kind: DataVolumeModel.kind,
+          metadata: {
+            name: claimName,
+            namespace,
+          },
+        },
+      }),
+    );
+};
+
+export const getMigrationClaimNameAnnotation = (migrationClaimName: string): string =>
+  `${ANNOTATION_PREFIX_MIGRATION_ORIGIN_CLAIMNAME}${migrationClaimName}`;
+
+export const createRollbackPatchData = (vm: V1VirtualMachine): Patch[] => {
+  const vmVolumes = getVolumes(vm);
+  const vmDataVolumeTemplates = getDataVolumeTemplates(vm);
+
+  const dataVolumeRollback = (vmDataVolumeTemplates || [])
+    .map((dataVolume, dvIndex) => {
+      const dataVolumeName = getName(dataVolume);
+
+      const volumeIndex = vmVolumes?.findIndex(
+        (volume) => dataVolumeName === volume?.dataVolume?.name,
+      );
+
+      const originalClaimAnnotation = getMigrationClaimNameAnnotation(dataVolumeName);
+
+      const originalClaimStorageMigration = getAnnotation(vm, originalClaimAnnotation);
+
+      if (!originalClaimStorageMigration) return [];
+
+      return [
+        {
+          op: 'remove',
+          path: `/metadata/annotations/${originalClaimAnnotation.replace('/', '~1')}`,
+        },
+        {
+          op: 'replace',
+          path: `/spec/template/spec/volumes/${volumeIndex}/dataVolume/name`,
+          value: originalClaimStorageMigration,
+        },
+        {
+          op: 'replace',
+          path: `/spec/dataVolumeTemplates/${dvIndex}/metadata/name`,
+          value: originalClaimStorageMigration,
+        },
+      ];
+    })
+    .filter(Boolean)
+    .flat();
+
+  const pvcRollback = (vmVolumes || [])
+    .filter((volume) => volume.persistentVolumeClaim?.claimName)
+    .map((volume, volumeIndex) => {
+      const pvcName = volume.persistentVolumeClaim.claimName;
+
+      const originalClaimAnnotation = getMigrationClaimNameAnnotation(pvcName);
+
+      const originalClaimStorageMigration = getAnnotation(vm, originalClaimAnnotation);
+
+      if (!originalClaimStorageMigration) return [];
+
+      return [
+        {
+          op: 'remove',
+          path: `/metadata/annotations/${originalClaimAnnotation.replace('/', '~1')}`,
+        },
+        {
+          op: 'replace',
+          path: `/spec/template/spec/volumes/${volumeIndex}/persistentVolumeClaim/claimName`,
+          value: originalClaimStorageMigration,
+        },
+      ];
+    })
+    .filter(Boolean)
```

Comment:
why u need filter here? seems like it always be true

Reply:
fixed with `reduce`